### PR TITLE
Check cnf claim with CSR fingerprint

### LINKED ISSUE #1637: Option to bind JWK provisioner tokens with CSR
### Description

In https://github.com/smallstep/certificates/discussions/1636, @daFritz84 proposes that binding JWT tokens for the JWK provisioner to a given CSR might be a good option to prevent some attacks.

Optionally, adding a new field in the token with a CSR fingerprint can be a good idea. Commands that generate their own token and CSR can always add the new field, and the JWK provisioner will validate it. `step ca token` can have an extra flag to add that fingerprint and a flag to pass a CSR to get the fingerprint from.

It might also be a good idea to add a fingerprint to `x509util.CertificateRequest`, so it can be optionally checked in a template or in a webhook, although the raw CSR is also present in a webhook request.

### PR DESCRIPTION
**Pull request by maraino**
This commit allows tying tokens with the provided  CSR or SSH public key. Tokens with a confirmation claim kid (`cnf.kid`) will validate that the provided fingerprint (kid) matches the CSR or SSH public key.

This check will only be present in JWK and X5C provisioners.

Fixes #1637

Related PRs:
- [x] https://github.com/smallstep/cli/pull/1092

### CODE REVIEW

File: authority/provisioner/sign_ssh_options.go [hunk 1]
```diff
@@ -419,6 +429,24 @@ func (v *sshNamePolicyValidator) Valid(cert *ssh.Certificate, _ SignSSHOptions)
 	}
 }
 
+// sshFingerprintValidator is a SSHPublicKeyValidator that checks the
+// fingerprint of the public key with the provided one.
+type sshFingerprintValidator string
+
+func (s sshFingerprintValidator) Valid(key ssh.PublicKey) error {
+	if s != "" {
```
[the remark this answers is not in the document]
Comment (maraino) [reply authority/provisioner/sign_ssh_options.go:437]: @hslatman I'm ok removing it for now, but with templates an X.509 CSR is just a signed public key. We do validate some of the fields there, but we could not do it.

[the remark this answers is not in the document]
Comment (maraino) [reply authority/provisioner/sign_ssh_options.go]: Fixed with 6c6ed46f

File: authority/provisioner/sign_options.go [hunk 2]
```diff
@@ -492,3 +495,21 @@ func (o *provisionerExtensionOption) Modify(cert *x509.Certificate, _ SignOption
 	cert.ExtraExtensions = append(cert.ExtraExtensions, ext)
 	return nil
 }
+
+// fingerprintValidator is a CertificateRequestValidator that checks the
+// fingerprint of the certificate with the provided one.
+type fingerprintValidator string
```
[the remark this answers is not in the document]
Comment (maraino) [reply authority/provisioner/sign_options.go]: Fixed with 6c6ed46f

File: authority/provisioner/x5c_test.go [hunk 3]
```diff
@@ -456,13 +458,36 @@ func TestX5C_AuthorizeSign(t *testing.T) {
...
+			}
+		},
 	}
 	for name, tt := range tests {
 		t.Run(name, func(t *testing.T) {
 			tc := tt(t)
 			ctx := NewContextWithMethod(context.Background(), SignIdentityMethod)
 			if opts, err := tc.p.AuthorizeSign(ctx, tc.token); err != nil {
-				if assert.NotNil(t, tc.err) {
+				if assert.NotNil(t, tc.err, err.Error()) {
```
Comment (hslatman) [inline authority/provisioner/x5c_test.go:490]: A bit of a weird test case. The `tc.err` is the expected error, and an assertion on that is unexpected. It was already there, but I don't think this is the right test logic now.

File: authority/tls_test.go [hunk 4]
```diff
@@ -249,6 +241,11 @@ func (e *testEnforcer) Enforce(cert *x509.Certificate) error {
 	return nil
 }
 
+func assertHasPrefix(t *testing.T, s, p string) bool {
+	t.Helper()
+	return assert.True(t, strings.HasPrefix(s, p), "%q is not a prefix of %q", p, s)
+}
```
Comment (hslatman) [inline authority/tls_test.go:247]: 👍

File: authority/provisioner/x5c_test.go (hunk 3 above)
Comment (maraino) [reply authority/provisioner/x5c_test.go:497]: It works as expected, showing the `err.Error()` if the assertion fails, but with ad70982c I've changed this file to use testify's assert and require packages.

**Review by hslatman — DISMISSED** [review]

**Review by hslatman — APPROVED** [review]

**Pull request merged**